#Fungal microbiomes: strategies for improving taxonomy resolution of environmental sequences

#Table of Contents
* [Authors]
* [Abstract]
* [Study Purpose]
* [Script Purpose]
* [System Requirements]

#Authors
* "Kristi Gdanetz MacCready"
* "Natalie Vande Pol"
* "Gregory Bonito"
* "Gian Maria NiccolÃ² Benucci"

#Abstract
>One of the most crucial steps in high-throughput sequence based microbiome studies is the taxonomic assignment of sequences belonging to operational taxonomic units (OTUs). Without taxonomic classification, functional and biological information of microbial communities cannot be inferred or interpreted. The Internal Transcribed Spacer (ITS) region of the nuclear ribosomal DNA (rDNA) is the conventional marker region for fungal community studies. While bioinformatics pipelines that cluster reads into OTUs have received much attention in the literature, the taxonomic classification of these sequences, upon which biological inference is dependent, has been largely neglected. Here we compared how the most common fungal OTU taxonomic assignment approaches, UTAX, RDP standalone, and SINTAX, handle ITS1 and ITS2 fungal sequence data generated by Illumina MiSeq amplicon sequencing. The classification power (ability to assign taxonomy at a given taxonomic rank) varied among the classifiers. Our results demonstrate how taxonomy assignment tools can bias understanding of fungal communities, and distort the linkage between taxa and specific treatments or ecosystems. We conclude with general guidelines and best practices for assigning taxonomy in fungal microbiome studies.

#Study Purpose

#Script Purpose

#System Requirements
